### PR TITLE
Filter related content by content type

### PR DESCRIPTION
**Pull request by RandyCupic**
Being able to filter tagged Content by Type is usefull when using Tags for relations, in that case it can happen that one side of relation holds many more Content items than the other side, and one is forced to paginate through the complete result set in order to find the Content.

This PR introduces content type filtering where one or more content types can be selected to filter the related content. It uses faceting to display only those content types that are available for this specific tag.

### CODE REVIEW

File: bundle/API/Repository/TagsService.php [hunk 1]
```diff
@@ -173,25 +173,40 @@ public function getTagSynonymCount(Tag $tag, array $languages = null, $useAlways
      * @param int $offset The start offset for paging
      * @param int $limit The number of content objects returned. If $limit = -1 all content objects starting at $offset are returned
      * @param bool $returnContentInfo
+     * @param array $sortClauses
```
Comment (emodric) [inline bundle/API/Repository/TagsService.php]: Can we use `\eZ\Publish\API\Repository\Values\Content\Query\SortClause[]` as a typehint here? Same goes for this change in `Core\TagsService` and `SignalSlot\TagsService` classes.

File: bundle/API/Repository/TagsService.php [hunk 2]
```diff
@@ -173,25 +173,40 @@ public function getTagSynonymCount(Tag $tag, array $languages = null, $useAlways
...
+     * @param \eZ\Publish\API\Repository\Values\Content\Query\Criterion[] $additionalCriteria Additional criteria for filtering related content
      *
-     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException If the current user is not allowed to read tags
-     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException If the specified tag is not found
+     * @throws \eZ\Publish\Core\Base\Exceptions\UnauthorizedException
```
Comment (emodric) [inline bundle/API/Repository/TagsService.php]: Why the change in `@throws` tags here? Same goes for this change in `Core\TagsService` and `SignalSlot\TagsService` classes.

File: bundle/API/Repository/TagsService.php [hunk 3]
```diff
@@ -173,25 +173,40 @@ public function getTagSynonymCount(Tag $tag, array $languages = null, $useAlways
...
+     * for content tagged with $tag.
+     *
+     * @param \Netgen\TagsBundle\API\Repository\Values\Tags\Tag $tag
+     * @param \eZ\Publish\API\Repository\Values\Content\Query\FacetBuilder[] $facetBuilders
+     *
+     * @throws \Netgen\TagsBundle\Exception\FacetingNotSupportedException
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Search\Facet[]
+     */
+    public function getRelatedContentFacets(Tag $tag, array $facetBuilders = []);
```
Comment (emodric) [inline bundle/API/Repository/TagsService.php]: Maybe this should be extracted to a separate service, since this method is not really useful by itself to warrant it being placed in public API. What do you think?

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php [hunk 4]
```diff
@@ -54,9 +64,30 @@ public function setTag(Tag $tag)
...
+     */
+    public function setSortClauses(array $sortClauses)
+    {
+        $this->sortClauses = $sortClauses;
+    }
+
+    /**
+     * Sets additional criteria to be used in search.
+     *
+     * @param array $additionalCriteria
```
Comment (emodric) [inline bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: Typehint should be `\eZ\Publish\API\Repository\Values\Content\Query\Criterion[]`, right?

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php [hunk 5]
```diff
@@ -54,9 +64,30 @@ public function setTag(Tag $tag)
...
+     */
+    public function setAdditionalCriteria(array $additionalCriteria = [])
+    {
+        $this->additionalCriteria = $additionalCriteria;
+    }
+
     /**
      * Returns the number of results.
      *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
```
Comment (emodric) [inline bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: We don't need these here, they're documented on tags service. Besides, this specific method does not throw exceptions by itself.

Same goes for `getSlice` method below.

File: bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace Netgen\TagsBundle\API\Repository\Values\Content\Query\SortClause;
+
+use eZ\Publish\API\Repository\Values\Content\Query;
+use eZ\Publish\API\Repository\Values\Content\Query\SortClause;
+
+/**
+ * Sets sort direction on Content Type ID for a content query.
+ */
+class ContentTypeId extends SortClause
```
Comment (emodric) [inline bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php]: Would content type identifier be a better sort clause than the ID? It makes more sense to sort by the identifier than a numeric database ID, right?

File: bundle/Core/Search/Solr/Query/Common/SortClauseVisitor/ContentTypeId.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Visits the sortClause tree into a Solr query.
+ */
+class ContentTypeId extends SortClauseVisitor
+{
+    /**
+     * Check if visitor is applicable to current sortClause.
+     *
+     * @param SortClause $sortClause
```
Comment (emodric) [inline bundle/Core/Search/Solr/Query/Common/SortClauseVisitor/ContentTypeId.php]: FQCN should be used in PHPDoc `@param` tag, here and in method below.

File: bundle/Form/Type/RelatedContentFilterType.php [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public function __construct(TagsService $tagsService, ContentTypeService $contentTypeService, SortService $sortService)
+    {
+        $this->tagsService = $tagsService;
+        $this->contentTypeService = $contentTypeService;
+        $this->sortService = $sortService;
+    }
+
+    /**
+     * {@inheritdoc}
```
Comment (emodric) [inline bundle/Form/Type/RelatedContentFilterType.php]: No need for `@inheritdoc` here. PHPDocs are inherited by default.

File: bundle/Form/Type/RelatedContentFilterType.php [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+
+        $resolver
+            ->setDefault('tag', null)
```
Comment (emodric) [inline bundle/Form/Type/RelatedContentFilterType.php]: `setDefault` is superfluous here. The value is always `null` by default. Besides, `null` is not in the list of allowed types two lines after this (which is fine).

File: bundle/Core/Search/RelatedContent/SortService.php [hunk 10]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Netgen\TagsBundle\Core\Search\RelatedContent;
+
+use eZ\Publish\API\Repository\Values\Content\Query;
+use eZ\Publish\API\Repository\Values\Content\Query\SortClause;
+use Netgen\TagsBundle\API\Repository\Values\Content\Query\SortClause\ContentTypeId;
+
+class SortService
```
Comment (emodric) [inline bundle/Core/Search/RelatedContent/SortService.php]: Probably a better name for this would be `SortClauseMapper` :)

File: bundle/API/Repository/TagsService.php (hunk 1 above)
Comment (RandyCupic) [reply bundle/API/Repository/TagsService.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/API/Repository/TagsService.php (hunk 2 above)
Comment (RandyCupic) [reply bundle/API/Repository/TagsService.php]: It was probably due to PHP Storm complaining about never thrown exception. Fixed in . But to be honest, I don't see where does NotFoundException come from? SearchService doesn't throw it.

Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php (hunk 4 above)
Comment (RandyCupic) [reply bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php (hunk 5 above)
Comment (RandyCupic) [reply bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: It was due to PHP Storm complaining because the exception is being thrown from TagsService but not being caught in this method so, in theory, `getNbResults` and `getSlice` could throw it as well? But if you think that it's better with it, I've fixed it in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/Core/Search/Solr/Query/Common/SortClauseVisitor/ContentTypeId.php (hunk 7 above)
Comment (RandyCupic) [reply bundle/Core/Search/Solr/Query/Common/SortClauseVisitor/ContentTypeId.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/Form/Type/RelatedContentFilterType.php (hunk 8 above)
Comment (RandyCupic) [reply bundle/Form/Type/RelatedContentFilterType.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/05f7dad27c86fbfdc247c7f08f2966bc63b3a183.

File: bundle/Form/Type/RelatedContentFilterType.php (hunk 9 above)
Comment (RandyCupic) [reply bundle/Form/Type/RelatedContentFilterType.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/2b99f2d4810b0d86e127859ae9364dd0f2f1921a.

File: bundle/Core/Search/RelatedContent/SortService.php (hunk 10 above)
Comment (RandyCupic) [reply bundle/Core/Search/RelatedContent/SortService.php]: I named it `service` because it doesn't only map sort clauses but it also returns valid sort options (to be used in form dropdown element) so it's kind of a all-in-one sort service. Should I rename it anyways?

File: bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php (hunk 6 above)
Comment (RandyCupic) [reply bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php]: Yes it would, but identifier is not being indexed in Solr.

File: bundle/API/Repository/TagsService.php (hunk 3 above)
Comment (RandyCupic) [reply bundle/API/Repository/TagsService.php]: @emodric I was struggling with that as well but in the absence of a better idea I've implemented it in the easiest way. Maybe a separate `RelatedContentService` would do?

File: bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php (hunk 6 above)
Comment (emodric) [reply bundle/API/Repository/Values/Content/Query/SortClause/ContentTypeId.php]: Really? That's unfortunate...

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php (hunk 5 above)
Comment (emodric) [reply bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: Yeah, PHPstorm needs to be configured correctly. Not every thrown exception in the service needs to be caught somewhere.

File: bundle/API/Repository/TagsService.php (hunk 3 above)
Comment (emodric) [reply bundle/API/Repository/TagsService.php]: Why not just `RelatedContentFacetBuilder` since it builds facets for related content search? `Service` has a different semantics in the domain.

File: bundle/Core/Search/RelatedContent/SortService.php (hunk 10 above)
Comment (emodric) [reply bundle/Core/Search/RelatedContent/SortService.php]: Yes, please. To me, it doesn't matter where it is used, it's what it does. And in essence, it maps sort clauses.

File: bundle/API/Repository/TagsService.php (hunk 2 above)
Comment (emodric) [reply bundle/API/Repository/TagsService.php]: This is a good catch indeed. `NotFoundException` is not thrown here, but I'd rather we fix it in one go later when cleaning up for v4.

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php [hunk 11]
```diff
@@ -78,6 +108,10 @@ public function getNbResults()
      * @param int $offset The offset
      * @param int $length The length
      *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
```
Comment (emodric) [inline bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: Also superfluous.

File: bundle/API/Repository/TagsService.php (hunk 2 above)
Comment (RandyCupic) [reply bundle/API/Repository/TagsService.php]: Fine with me :)

File: bundle/Core/Search/RelatedContent/SortService.php (hunk 10 above)
Comment (RandyCupic) [reply bundle/Core/Search/RelatedContent/SortService.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/45691e2549bde8a5056a3fa8b1a80aca28825848.

File: bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php (hunk 11 above)
Comment (RandyCupic) [reply bundle/Core/Pagination/Pagerfanta/RelatedContentAdapter.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/f323b02f1c1dbcedffd2e2c09268ba3d07ac8f5f.

File: bundle/API/Repository/TagsService.php (hunk 3 above)
Comment (RandyCupic) [reply bundle/API/Repository/TagsService.php]: Fixed in https://github.com/netgen/TagsBundle/pull/98/commits/46986d8987d499d61b6cc3efb128e37fed405940.

**Comment by emodric** [comment]
Thanks @RandyCupic !
